File: go/math/50_my_pow.go
package math

// myPow Pow(x, n)
func myPow(x float64, n int) float64 {
	// 快速幂法
	if n > 0 {
		return quickMu(x, n)
	}

	return 1 / quickMu(x, -n)
}

func quickMu(x float64, n int) float64 {

	if n == 0 {
		return 1
	}

	y := quickMu(x, n/2)

	// 偶数
	if n%2 == 0 {
		return y * y
	}
	// 奇数
	return y * y * x
}

File: rust/src/math/_50_my_pow.rs
// Pow(x, n)
pub fn my_pow(x: f64, n: i32) -> f64 {
    // 快速幂法
    if n >= 0 {
        return quick_mu(x, n);
    }

    1.0 / quick_mu(x, -n)
}

pub fn quick_mu(x: f64, n: i32) -> f64 {
    if n == 0 {
        return 1.0;
    }

    let y = quick_mu(x, n / 2);

    if n % 2 == 0 {
        return y * y;
    }
    y * y * x
}
